src/utils/index.py

```diff
@@ -0,0 +1,283 @@
+import os
+import logging
+import argparse
+import json
+from typing import Set, Iterator, Tuple, List, Optional
+from pathlib import Path
+from lib.ursadb import UrsaDb
+from multiprocessing import Pool
+
+
+def all_indexed_files(ursa: UrsaDb) -> Set[str]:
+    iterator = ursa.query("{}")["iterator"]
+    result: Set[str] = set()
+    while True:
+        pop_result = ursa.pop(iterator, 10000)
+        if pop_result.iterator_empty:
+            break
+        for fpath in pop_result.files:
+            result.add(fpath)
+    return result
+
+
+def walk_directory(dir: Path, ignores: List[str]) -> Iterator[Path]:
+    """Recursively walks the current directory, while respecting .ursadbignore
+    files to selectively ignore some elements """
+    if (dir / ".ursadb").exists():
+        new_config = (dir / ".ursadb").read_text().strip().split("\n")
+        for line in new_config:
+            if line.startswith("ignore:"):
+                ignores.append(line[len("ignore:") :].strip())
+    for elem in dir.iterdir():
+        if any(elem.match(ignore) for ignore in ignores):
+            continue
+        elif elem.is_file():
+            yield elem
+        elif elem.is_dir():
+            for elem in walk_directory(elem, ignores):
+                yield elem
+
+
+def find_new_files(
+    existing: Set[str], files_root: Path, mounted_as: str
+) -> Iterator[str]:
+    for abspath in walk_directory(files_root.resolve(), [".ursadb"]):
+        relpath = os.path.relpath(abspath, files_root)
+        mounted_path = os.path.join(mounted_as, relpath)
+        if mounted_path not in existing:
+            yield str(abspath)
+
+
+def index_files(
+    proc_params: Tuple[str, List[str], Path, Optional[Tuple[str, str]]]
+) -> str:
+    ursa_url, types, batch, path_rel = proc_params
+    ursa = UrsaDb(ursa_url)
+    with_ = ", ".join(types)
```

Comment:
Maybe it's better just to name it `types_` to make it easier to track the original argument.